File: app/src/main/java/com/example/ugcssample/activities/MainActivity.java
package com.example.ugcssample.activities;

import android.annotation.SuppressLint;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.ServiceConnection;
import android.content.SharedPreferences;
import android.os.Build;
import android.os.Bundle;
import android.os.IBinder;
import android.preference.PreferenceManager;
import android.view.View;
import android.widget.Button;
import android.widget.ImageButton;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import com.example.ugcssample.R;
import com.example.ugcssample.drone.DroneBridgeImpl;
import com.example.ugcssample.fragment.VideoViewFragment;
import com.example.ugcssample.services.DjiAppMainService;
import com.example.ugcssample.services.DjiAppMainServiceBinder;
import com.example.ugcssample.services.DjiAppMainServiceImpl;
import com.example.ugcssample.utils.ArrayUtils;
import com.example.ugcssample.utils.PermissionUtils;

import java.util.List;

import dji.sdk.camera.VideoFeeder;
import dji.ux.widget.FPVOverlayWidget;
import dji.ux.widget.FPVWidget;
import timber.log.Timber;

public class MainActivity extends AppCompatActivity {

    private static final IntentFilter EVENT_FILTER = new IntentFilter();

    static {
        EVENT_FILTER.addAction(DroneBridgeImpl.ON_DRONE_CONNECTED);
    }
    private Intent sIntent;
    private ServiceConnection sConn;
    private Button btnSimulator;
    private Button bindRc;
    private Button wideButton;
    private Button zoomButton;
    private Button bindCustomMedia;
    private Button bindWidgets;
    private FPVWidget fpvWidget;
    private FPVOverlayWidget fpvOverlayWidget;
    private FPVWidget secondaryFpvWidget;
    private boolean switched = false;
    private ImageButton btnPreferences;
    protected DjiAppMainService appMainService;
    public static final int REQUEST_PERMISSION_CODE = 2358;
    LocalBroadcastManager broadcastManager;
  //  private VideoViewFragment primaryVideoFeedView;

    private final BroadcastReceiver eventReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {

            final String action = intent.getAction();
            if (action == null)
                return;

            if (DroneBridgeImpl.ON_DRONE_CONNECTED.equals(action)) {
               // btnSimulator.setEnabled(true);
                bindRc.setEnabled(true);
              //  bindCustomMedia.setEnabled(true);
                bindWidgets.setEnabled(true);
                wideButton.setEnabled(true);
                zoomButton.setEnabled(true);
               // primaryVideoFeedView.registerLiveVideo(VideoFeeder.getInstance().getPrimaryVideoFeed(), true);
            }
        }
    };


    @Override
    @SuppressLint("CommitPrefEdits")
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);


        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(this);
        String list_preference_C1 = prefs.getString("list_preference_C1", "");
        String list_preference_C2 = prefs.getString("list_preference_C2", "");
        String list_preference_C1_C2_release = prefs.getString("list_preference_C1_C2_release", "");
        String list_preference_5d_left = prefs.getString("list_preference_5d_left", "");
        String list_preference_5d_right = prefs.getString("list_preference_5d_right", "");
        String list_preference_5d_up = prefs.getString("list_preference_5d_up", "");
        String list_preference_5d_down = prefs.getString("list_preference_5d_down", "");
        String list_preference_5d_push = prefs.getString("list_preference_5d_push", "");
        String list_preference_5d_middle = prefs.getString("list_preference_5d_middle", "");

       // <item>ZoomInContinue</item>
       // <item>ZoomOutContinue</item>
      //  <item>ZoomStop</item>
      //  <item>CamUp</item>
      //  <item>CamDown</item>
      //  <item>CamLeft</item>
      //  <item>CamRight</item>
       // <item>CamStop</item>

        SharedPreferences.Editor editor = prefs.edit();
        if (list_preference_C1.equals("")) {
            editor.putString("list_preference_C1", "ZoomInContinue");
            editor.commit();
        }
        if (list_preference_C2.equals("")) {
            editor.putString("list_preference_C2", "ZoomOutContinue");
            editor.commit();
        }
        if (list_preference_C1_C2_release.equals("")) {
            editor.putString("list_preference_C1_C2_release", "ZoomStop");
            editor.commit();
        }
        if (list_preference_5d_left.equals("")) {
            editor.putString("list_preference_5d_left", "CamLeft");
            editor.commit();
        }
        if (list_preference_5d_right.equals("")) {
            editor.putString("list_preference_5d_right", "CamRight");
            editor.commit();
        }
        if (list_preference_5d_up.equals("")) {
            editor.putString("list_preference_5d_up", "CamUp");
            editor.commit();
        }
        if (list_preference_5d_down.equals("")) {
            editor.putString("list_preference_5d_down", "CamDown");
            editor.commit();
        }
        if (list_preference_5d_push.equals("")) {
            editor.putString("list_preference_5d_push", "CamResetToMiddle");
            editor.commit();
        }
        if (list_preference_5d_middle.equals("")) {
            editor.putString("list_preference_5d_middle", "CamStop");
            editor.commit();
        }

        sIntent = new Intent(this, DjiAppMainServiceImpl.class);
        sConn = new ServiceConnection() {
            public void onServiceConnected(ComponentName name, IBinder binder) {
                MainActivity.this.onMainServiceConnected(name, binder);
            }

            public void onServiceDisconnected(ComponentName name) {
                MainActivity.this.onMainServiceDisconnected();
            }
        };

     //   primaryVideoFeedView = (VideoViewFragment) findViewById(R.id.video_view_primary_video_feed);
      //  btnSimulator = (Button) findViewById(R.id.btn_simulator);
      //  btnSimulator.setOnClickListener(v -> {
      //      appMainService.startSimulator();
      //  });

        wideButton = (Button) findViewById(R.id.wide);
        wideButton.setOnClickListener(v -> {
            appMainService.wide();
        });


        zoomButton = (Button) findViewById(R.id.zoom);
        zoomButton.setOnClickListener(v -> {
            appMainService.zoom();
        });

        bindRc = (Button) findViewById(R.id.btn_rc);
        bindRc.setOnClickListener(v -> {
            appMainService.bindRC();
        });
       // bindCustomMedia = (Button) findViewById(R.id.btn_custom_media);
       // bindCustomMedia.setOnClickListener(v -> {
       //     appMainService.setMedia("Custom media info");
       // });

        fpvWidget = (FPVWidget)findViewById(R.id.fpv_custom_widget);
        fpvOverlayWidget = (FPVOverlayWidget) findViewById(R.id.fpv_overlay_widget);
        secondaryFpvWidget = (FPVWidget)findViewById(R.id.secondary_fpv_custom_widget);

        bindWidgets = (Button) findViewById(R.id.switch_button_primary);
        bindWidgets.setOnClickListener(v -> {
            if (!switched) {
                fpvWidget.setVideoSource(FPVWidget.VideoSource.SECONDARY);
                secondaryFpvWidget.setVideoSource(FPVWidget.VideoSource.PRIMARY);
            } else {
                fpvWidget.setVideoSource(FPVWidget.VideoSource.PRIMARY);
                secondaryFpvWidget.setVideoSource(FPVWidget.VideoSource.SECONDARY);
            }
            switched = !switched;
        });

        btnPreferences = findViewById(R.id.btn_preferences);
        btnPreferences.setOnClickListener((v -> {
            Intent intent = new Intent(this, PreferencesActivity.class);
            startActivity(intent);
        }));
    }

    protected void onMainServiceConnected(ComponentName name, IBinder binder) {
        appMainService = ((DjiAppMainServiceBinder)binder).getService();
        checkAndRequestAndroidPermissions();
    }

    protected void onMainServiceDisconnected() {
    }

    @Override
    protected void onResume() {
        super.onResume();
        bindService(sIntent, sConn, BIND_AUTO_CREATE);
        broadcastManager = LocalBroadcastManager.getInstance(this);
        broadcastManager.registerReceiver(eventReceiver, EVENT_FILTER);
    }

    @Override
    protected void onPause() {
        super.onPause();
    }

    @Override
    protected void onDestroy() {
        unbindService(sConn);
        super.onDestroy();
    }

    private void checkAndRequestAndroidPermissions() {
        List<String> missing = PermissionUtils.checkForMissingPermission(getApplicationContext());
        if (!ArrayUtils.isEmpty(missing)) {
            String[] missingPermissions = missing.toArray(new String[0]);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                ActivityCompat.requestPermissions(this, missingPermissions, REQUEST_PERMISSION_CODE);
            } else {
                //failed
            }
        } else {
            //all android permissions granted;
            onAndroidPermissionsValid();
        }
    }
    protected void onAndroidPermissionsValid() {
        appMainService.init();
    }
    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        // Check for granted permission and remove from missing list
        if (requestCode == REQUEST_PERMISSION_CODE) {
            Timber.d("onRequestPermissionsResult...");
            checkAndRequestAndroidPermissions();
        }
    }
}

File: app/src/main/java/com/example/ugcssample/activities/MainActivity.kt
package com.example.ugcssample.activities

import android.content.*
import android.os.Build
import android.os.Bundle
import android.os.IBinder
import android.view.View
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.example.ugcssample.R
import com.example.ugcssample.drone.DroneBridgeImpl
import com.example.ugcssample.fragment.VideoViewFragment
import com.example.ugcssample.services.DjiAppMainService
import com.example.ugcssample.services.DjiAppMainServiceBinder
import com.example.ugcssample.services.DjiAppMainServiceImpl
import com.example.ugcssample.utils.ArrayUtils
import com.example.ugcssample.utils.ExceptionWriter
import com.example.ugcssample.utils.PermissionUtils
import dji.sdk.camera.VideoFeeder
import timber.log.Timber

class MainActivity : AppCompatActivity() {
    companion object {
        private val EVENT_FILTER = IntentFilter()
        const val REQUEST_PERMISSION_CODE = 2358

        init {
            DroneBridgeImpl.DroneActions.values().forEach { EVENT_FILTER.addAction(it.name) }
            
        }
    }

    private var sIntent: Intent? = null
    private var sConn: ServiceConnection? = null
    private var btnDetectCameraModes: Button? = null
    private var tvStatus: TextView? = null
    protected var appMainService: DjiAppMainService? = null
    var broadcastManager: LocalBroadcastManager? = null
    private var primaryVideoFeedView: VideoViewFragment? = null
    private val eventReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context : Context, intent : Intent) {
            val action = intent.action ?: return
            when (DroneBridgeImpl.DroneActions.valueOf(action)) {
                DroneBridgeImpl.DroneActions.ON_DRONE_CONNECTED ->
                {
                    tvStatus?.text = ("Drone ready. Press 'Start Tests'")
                    btnDetectCameraModes!!.isEnabled = true
                    primaryVideoFeedView!!.registerLiveVideo(
                            VideoFeeder.getInstance().primaryVideoFeed,
                            true
                                                            )
                }
                DroneBridgeImpl.DroneActions.ON_DRONE_DISCONNECTED ->
                {
                    tvStatus?.text = ("Drone disconnected. Connect drone")
                }
                DroneBridgeImpl.DroneActions.CAMERA_TESTS_STARTED ->
                {
                    btnDetectCameraModes?.isEnabled = false
                    tvStatus?.text = ("Tests started. Please, wait")
                    Toast.makeText(context, "Tests started. Please, wait", Toast.LENGTH_SHORT).show()
                }
                DroneBridgeImpl.DroneActions.CAMERA_TESTS_PROGRESS ->
                {
                    tvStatus?.text = ("Tests running. Test #${intent.getIntExtra("test-index", -1)}")
                }
                DroneBridgeImpl.DroneActions.CAMERA_TESTS_FINISHED ->
                {
                    btnDetectCameraModes?.isEnabled = true
                    tvStatus?.text = ("File dumped at\n${intent.getStringExtra("file-path")}")
                    Toast.makeText(context, "Tests finished. File dumped", Toast.LENGTH_SHORT).show()
                }
            }
        }
    }

    override fun onCreate(savedInstanceState : Bundle?) {
        ExceptionWriter.setupCrashHandler(applicationContext)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        sIntent = Intent(this, DjiAppMainServiceImpl::class.java)
        sConn = object : ServiceConnection {
            override fun onServiceConnected(name : ComponentName, binder : IBinder) {
                onMainServiceConnected(name, binder)
            }

            override fun onServiceDisconnected(name : ComponentName) {
                onMainServiceDisconnected()
            }
        }
        primaryVideoFeedView =
            findViewById<View>(R.id.video_view_primary_video_feed) as VideoViewFragment
        btnDetectCameraModes = findViewById(R.id.btn_detect_camera_modes)
        tvStatus = findViewById(R.id.tv_status)
        btnDetectCameraModes?.setOnClickListener {
            appMainService!!.testCameraModes()
            
        }
    
        
    }

    protected fun onMainServiceConnected(name : ComponentName?, binder : IBinder) {
        appMainService = (binder as DjiAppMainServiceBinder).getService()
        checkAndRequestAndroidPermissions()
    }

    protected fun onMainServiceDisconnected() {}
    override fun onResume() {
        super.onResume()
        bindService(sIntent, sConn!!, BIND_AUTO_CREATE)
        broadcastManager = LocalBroadcastManager.getInstance(this)
        broadcastManager!!.registerReceiver(eventReceiver, EVENT_FILTER)
    }

    override fun onPause() {
        super.onPause()
    }

    override fun onDestroy() {
        unbindService(sConn!!)
        super.onDestroy()
    }

    private fun checkAndRequestAndroidPermissions() {
        val missing = PermissionUtils.checkForMissingPermission(applicationContext)
        if (!ArrayUtils.isEmpty(missing)) {
            val missingPermissions = missing?.toTypedArray()
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && missingPermissions != null) {
                ActivityCompat.requestPermissions(this, missingPermissions, REQUEST_PERMISSION_CODE)
            } else {
                //failed
            }
        } else {
            //all android permissions granted;
            onAndroidPermissionsValid()
        }
    }

    protected fun onAndroidPermissionsValid() {
        appMainService!!.init()
    }

    override fun onRequestPermissionsResult(
            requestCode : Int,
            permissions : Array<String>,
            grantResults : IntArray
                                           ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        // Check for granted permission and remove from missing list
        if (requestCode == REQUEST_PERMISSION_CODE) {
            Timber.d("onRequestPermissionsResult...")
            checkAndRequestAndroidPermissions()
        }
    }
}